### PR TITLE
Implement custom dynamic headers, fix bug where background tasks never end in certain situations

### PR DESCRIPTION
**Pull request by marvukusic**
**Implement custom dynamic headers**
- Custom dynamic headers are added as TUSUpload init parameters
- Purpose is to enable different header for every file when multiple files are added to upload queue
- Existing custom headers when creating or resuming an upload are not adequate to be dynamic as they are saved in TUSExecutor which exists only as an instance in TUSClient singleton. The result is that only last set customHeader parameter is used

**Fix bug where background tasks never end in certain situations**
- When multiple uploads are in queue endBackgroundTask function is called only when last one finishes
- Result is that background tasks of first n-1 uploads are never ended

### CODE REVIEW

File: TUSKit/Classes/TUSExecutor.swift
Comment (hannojg) [inline TUSKit/Classes/TUSExecutor.swift:218]: hmm, I am unsure about this change. Calling completion will stop the background task. So, even when we are calling `continueUploading` the background task will stop.
This code:
```
            let pendingUploads = TUSClient.shared.pendingUploads()
            TUSClient.shared.createOrResume(forUpload: pendingUploads[0])
```
is being run in the background task we just stopped 🤔 
Or am I missing something?

**Review by hannojg — CHANGES_REQUESTED** [review]

File: TUSKit/Classes/TUSExecutor.swift
Comment (marvukusic) [reply TUSKit/Classes/TUSExecutor.swift:218]: That change is in `handleUploadSuccess(upload:completion:)` method which means that upload is completed and background task should be closed. 

In case there are more files to upload, `createOrResume(forUpload:)` method will be triggered, which opens new background task for that upload. This should work well in foreground state, but remains to be seen if this is possible when app goes to background and `TUSBackgroundMode.PreferUploadQueue` is set as `backgroundMode`. In my opinion `PreferUploadQueue` should be avoided in most situations as there are no guarantees that whole queue will be uploaded on time allocated by iOS.

As a proof for this issue, console warnings of still running background tasks can be observed in case when there is more than one file in upload queue. Number of background task warnings equals to `N-1`, where `N` is the number of files in queue. `N-1` is because only last file in queue closes its background task.

While we are at it, same situation is in markAsFailed(upload:completion:) method.

**Comment by dfmarulanda** [comment]
Any update on this?

**Comment by marvukusic** [comment]
> Any update on this?

From my side?
I don't have much to say on this matter. Currently, I'm using the library with proposed changes in my own fork and waiting for pull request approval, so I can get back to using it from the original source.

**Comment by Acconut** [comment]
Apologies for the delay! We are currently waiting for our new maintainer to begin working on TUSKit at the beginning of September. I am afraid we have to wait with a proper review until then as I don't know any Swift :)

**Comment by tjeerdintveen** [comment]
Hi,  I will be the new contributor ☺️ 
Thanks for making the PR.

I just started, but at this stage I am planning to rework the internals, so I can make a strong test suite. I will also dive deep into the background execution. That means I will take these changes into consideration, but I do expect a lot of changes, which might means this PR won't be needed after.

I do however, took a note of your requirements regarding custom headers.

I will keep you posted.

**Pull request closed without merging**
